cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java

```diff
@@ -0,0 +1,429 @@
+   * @param start start time (inclusive) in milliseconds
+   * @param end end time (exclusive) in milliseconds
+   * @return the operations and program run information
+   */
+  public Set<ProgramRunOperations> getOutgoingOperations(EndPoint endPoint, long start, long end) {
+    return getOperations(OUTGOING_DIRECTION_MARKER, endPoint, start, end);
+  }
+
+  private Set<ProgramRunOperations> getOperations(byte[] direction, EndPoint endPoint, long start, long end) {
+    Map<Long, Set<ProgramRunId>> checksumsWithProgramRunsInRange
+            = getChecksumsWithProgramRunsInRange(direction, endPoint, start, end);
+
+    Set<ProgramRunOperations> result = new HashSet<>();
+
+    for (Map.Entry<Long, Set<ProgramRunId>> entry : checksumsWithProgramRunsInRange.entrySet()) {
+      byte[] rowKey = getChecksumRowKey(entry.getKey());
+      String value = Bytes.toString(table.get(rowKey, RAW_OPERATION_MARKER));
+      Set<Operation> operations = null;
+      try {
+        operations = GSON.fromJson(value, SET_OPERATION_TYPE);
+      } catch (JsonSyntaxException e) {
+        LOG.warn(String.format("Failed to parse json from row '%s' and column '%s'.", Bytes.toString(rowKey),
+                Bytes.toString(RAW_OPERATION_MARKER)));
+      }
+      if (operations != null) {
+        result.add(new ProgramRunOperations(entry.getKey(), checksumsWithProgramRunsInRange.get(entry.getKey()),
+                                            operations));
+      }
+    }
+
+    return result;
+  }
+
+  private Map<Long, Set<ProgramRunId>> getChecksumsWithProgramRunsInRange(byte[] direction, EndPoint endPoint,
+                                                                          long start, long end) {
+    // time is inverted, hence we need to pass end-time for getting start key
+    byte[] scanStartKey = getScanKey(direction, endPoint, end);
+    // time is inverted, hence we need to pass start-time for getting end key
+    byte[] scanEndKey = getScanKey(direction, endPoint, start);
+    Map<Long, Set<ProgramRunId>> result = new HashMap<>();
+    try (Scanner scanner = table.scan(scanStartKey, scanEndKey)) {
+      Row row;
+      while ((row = scanner.next()) != null) {
+        if (LOG.isTraceEnabled()) {
+          LOG.trace(String.format("Got row key = '%s' for direction '%s', EndPoint '%s', and time range ('%s'-'%s').",
+                                  Bytes.toString(row.getRow()), Bytes.toString(direction), endPoint, start, end));
+        }
+
+        long checksum = Bytes.toLong(row.get(CHECKSUM_MARKER));
+        ProgramRunId programRunId = null;
+        try {
+          programRunId = GSON.fromJson(Bytes.toString(row.get(PROGRAM_MARKER)), ProgramRunId.class);
+        } catch (JsonSyntaxException e) {
+          LOG.warn(String.format("Failed to parse json from row '%s' and column '%s'.", Bytes.toString(row.getRow()),
+                  Bytes.toString(PROGRAM_MARKER)));
+        }
```

Comment:
better ```continue;``` here, then you don't need the ```if (programRunId != null)``` below.

Reply:
done